src/wallet/rpc/transactions.cpp

```diff
@@ -866,15 +866,17 @@ RPCHelpMan rescanblockchain()
     wallet.BlockUntilSyncedToCurrentChain();
 
     WalletRescanReserver reserver(*pwallet);
-    if (!reserver.reserve()) {
+    if (!reserver.reserve(/*with_passphrase=*/true)) {
         throw JSONRPCError(RPC_WALLET_ERROR, "Wallet is currently rescanning. Abort existing rescan or wait.");
     }
 
     int start_height = 0;
     std::optional<int> stop_height;
     uint256 start_block;
+
     {
         LOCK(pwallet->cs_wallet);
+        EnsureWalletIsUnlocked(*pwallet);
```

Comment:
if the wallet is a watch-only wallet, do we need to ensure this?

Reply:
No, I will make it so that here and in the other RPCs `EnsureWalletIsUnlocked` is only called if the wallet is not watch-only. Additionally, in that case `m_relock_mutex` doesn't need to be held, as long as it is not grabbed later in the RPC.

Reply:
Actually, the wallet will always be unlocked if private keys are disabled, because wallet encryption is not allowed in a watch-only wallet. See: https://github.com/bitcoin/bitcoin/blob/master/src/wallet/rpc/encrypt.cpp#L224